yarn-project/p2p/src/mem_pools/tx_pool/aztec_kv_tx_pool.ts

```diff
@@ -300,4 +342,181 @@ export class AztecKVTxPool implements TxPool {
+   * @param blockNumber - The block number of the mined block.
+   * @returns The total number of txs evicted from the pool.
+   *
+   * TODO: add logging for evicted txs?
+   */
+  private async evictInvalidTxsAfterMining(minedTxHashes: TxHash[], blockNumber: number): Promise<number> {
+    if (minedTxHashes.length === 0) {
+      return 0;
+    }
+
+    const pendingTxHashes = await this.getPendingTxHashes();
+    const minedTxs = await Promise.all(minedTxHashes.map(async hash => await this.getTxByHash(hash)));
+    const minedNullifiers = new Set<Fr>();
+    for (const tx of minedTxs) {
+      if (tx) {
+        const nullifiers = tx.data.getNonEmptyNullifiers();
+        nullifiers.forEach(nullifier => minedNullifiers.add(nullifier));
+      }
+    }
+    const pendingTxs = await Promise.all(
+      pendingTxHashes.map(async hash => ({
+        hash,
+        tx: await this.getTxByHash(hash),
+      })),
+    );
+
+    const txsToEvict: TxHash[] = [];
+    for (const { hash, tx } of pendingTxs) {
+      if (!tx) {
+        continue;
+      }
+
+      // Evict pending txs that share nullifiers with mined txs
+      const txNullifiers = tx.data.getNonEmptyNullifiers();
+      if (txNullifiers.some(nullifier => minedNullifiers.has(nullifier))) {
+        txsToEvict.push(hash);
+        continue;
+      }
+
+      // Evict pending txs with an insufficient fee payer balance
+      const db = this.#worldStateSynchronizer.getCommitted();
+      const gasTxValidator = new GasTxValidator(
+        new DatabasePublicStateSource(db),
+        ProtocolContractAddress.FeeJuice,
+        GasFees.empty(),
+      );
+      if ((await gasTxValidator.validateTxFee(tx)).result === 'invalid') {
+        txsToEvict.push(hash);
+        continue;
+      }
+
+      // Evict pending txs with a max block number less than or equal to the mined block
+      const maxBlockNumber = tx.data.rollupValidationRequests.maxBlockNumber;
+      if (maxBlockNumber.isSome && maxBlockNumber.value.toNumber() <= blockNumber) {
+        txsToEvict.push(hash);
+        continue;
+      }
+    }
+
+    if (txsToEvict.length > 0) {
+      await this.deleteTxs(txsToEvict, true);
+    }
+    return txsToEvict.length;
+  }
+
+  /**
+   * Evicts pending txs that no longer have valid archive roots from the pool after a reorg.
+   *
+   * @param txHashes - The tx hashes of the txs that were moved from mined to pending.
+   * @returns The total number of txs evicted from the pool.
+   */
+  private async evictInvalidTxsAfterReorg(txHashes: TxHash[]): Promise<number> {
+    if (txHashes.length === 0) {
+      return 0;
+    }
+
+    const db = this.#worldStateSynchronizer.getCommitted();
+    const archiveCache = new ArchiveCache(db);
+    const pendingTxHashes = await this.getPendingTxHashes();
+    const pendingTxs = await Promise.all(
+      pendingTxHashes.map(async hash => ({
+        hash,
+        archiveRoot: await (await this.getTxByHash(hash))?.data.constants.historicalHeader.hash(),
```

Comment:
Let's also keep an eye on the cost of computing header hashes. We may have to store them to avoid recomputing.